pyomo/core/base/component.py

```diff
@@ -732,17 +743,22 @@ def __str__(self):
         """Return a string with the component name and index"""
         return self.name
 
-    def to_string(self, ostream=None, verbose=None, precedence=0, labeler=None):
+    def to_string(self, verbose=None, labeler=None, smap=None, compute_values=False):
         """
-        Write the component name and index to a buffer,
+        Return a string representation of this component,
         applying the labeler if passed one.
         """
-        if ostream is None:
-            ostream = sys.stdout
+        if compute_values:
+            try:
+                return str(self())
+            except:
+                pass
+        if smap:
+            return smap.getSymbol(self, labeler)
         if labeler is not None:
-            ostream.write(labeler(self))
+            return labeler(self)
         else:
-            ostream.write(self.__str__())
+            return self.__str__()
```

Comment:
This change in the API needs to be prominently documented.  I agree with it, but it is a change.